### DISCUSSION TITLE
CartesianChartModelProducer size

### DISCUSSION THREAD

**Discussion by fieqapp** [patrykandpatrick/vico#713] [Questions]
First of all, thank you for such a great library!

In my app I have created a fun which draws the chart receiving just the CartesianChartModelProducer.
There are time I draw a Chart with two lines or just one line.
Also the size of each serie maught be different.
I'm wondering if there is some way to get the quantity of the series and the size of each serie from CartesianChartModelProducer object?

I checked all the discussions and so on but could not find anything related to this.
Regards.

**Comment by joey101** [comment]
in the docs, it says
https://patrykandpatrick.com/vico/api/2.0.0-alpha.19/core/com.patrykandpatrick.vico.core.cartesian.data/-cartesian-chart-model/#-948773955%2FProperties%2F186187275

if you do ```CartesianChartModelProducer.width``` 
> val [width](https://patrykandpatrick.com/vico/api/2.0.0-alpha.19/core/
> com.patrykandpatrick.vico.core.cartesian.data/-cartesian-chart-model/width.html): [Float](https://
> kotlinlang.org/api/latest/jvm/stdlib/kotlin/-float/index.html)

> Expresses the size of this [CartesianChartModel](https://patrykandpatrick.com/vico/api/2.0.0-alpha.19/core/
> com.patrykandpatrick.vico.core.cartesian.data/-cartesian-chart-model/index.html) in terms of the range of 
> the x values covered.

correct me if i am wrong.

But I do not think you can have different series sizes.

**Reply by fieqapp** [reply]
Hi,

Thanks for the reply, the links are not working but I guess it was a link to **CartesianChartModel** and not **CartesianChartModelProducer**.
It may also work if there is any way to get the Model or Models from the Producer. Is there any way to do so?

**Reply by patrickmichalik** [reply]
Thanks for your input, @joey101! As mentioned by @fieqapp, `width` is a property of `CartesianChartModel`, which `CartesianChartModelProducer` creates instances of. It’s worth adding that in most cases, using `ChartValues#xLength` is preferable, as its value is consistent with `AxisValueOverrider` range changes.

@fieqapp, regarding manually retrieving the `CartesianChartModel` from the `CartesianChartModelProducer` and using it to perform setup, this should never be done, so there’s no API that facilitates it. The reasons and the recommended approach are described [here](https://patrykandpatrick.com/vico/wiki/cartesian-charts/data#asynchrony-and-extras). The `CartesianChartModel` can certainly be obtained, but by other means. I’ll get back to you [below](https://github.com/patrykandpatrick/vico/discussions/713#discussioncomment-9513680) to discuss the optimal solution as soon as I can.

**Comment by patrickmichalik** [comment]
Hello, @fieqapp! `CartesianChartModelProducer` doesn’t expose such information and, seeing how the update system is structured, wouldn’t be an appropriate source thereof. Its sole purpose is `CartesianChartModel` generation. It is `LineCartesianLayerModel` that makes such details available. The chart’s `LineCartesianLayerModel` is included in its `CartesianChartModel`, which you can access by several means. Could you describe what the end goal is? What are you looking to use these values for?

**Reply by fieqapp** [reply]
@patrickmichalik Thanks for the reply and specially for the library!
1) **Serie's length**: The purpose is to adjust the labels spacing on the x-axis, so based on the series length, the spacing is changing in order to have always 4 labels on the x-axis.
2) **Series quantity**: Additionally, as the `CartesianChartModelProducer ` may have more than one data series the legend does not make sense is there is only one series and do make sense if there are more than one data series. Also, I wanted to adjust the colors based on this parameter.
Right now my function received this info as additional parameters, but I was wondering if there is any way to simplify my code as this info is already a part of the `CartesianChartModelProducer`
3) If there is a possibility to access the data series also would be helpful to get MAX and MIN values to adjust the axis.

I was thinking to send the series to my function as a List of Map and create the `CartesianChartModelProducer` in the function, but the documentation recommends to create the `CartesianChartModelProducer` in the viewmodel, that's why I create it in the viewmodel and then send to the function.

**Reply by patrickmichalik** [reply]
Apologies for the late reply, and thanks for the kind words!

As discussed in [subsection 2.2.4](https://www.patrykandpatrick.com/vico/guide/latest/core/cartesian-charts/cartesian-chart-model-producer#224-asynchrony-and-extras) of the guide, dynamic setup tied to `CartesianChartModel` updates must be performed based on Vico-provided arguments of lambdas, interface functions, and the like. This relates to asynchronous processing. External channels—such as the composable-parameter solution you mentioned—provide insufficient synchronization and should be avoided. Switching to the recommended approach will naturally achieve the desired result of having only a `CartesianChartModelProducer` parameter in the composable. More importantly, it will ensure reliable behavior during data updates.

1. Since you’ve not specified otherwise, I assume you’re using `HorizontalAxis.ItemPlacer.aligned` (for Vico 2.0.0 Beta 1 and later), `HorizontalAxis.ItemPlacer.default` (for Vico 2.0.0 Alpha 23 and later), or `AxisItemPLacer.Horizontal.default` (for older versions).

   Since Vico 2.0.0 Beta 4, the `spacing` parameter of `HorizontalAxis.ItemPlacer.aligned` is an `ExtraStore` lambda. Thus, if the desired label count is constant and accessible from the `ViewModel`, you can implement this behavior by introducing an extra that defines the desired spacing. Its value would be calculated in the `CartesianChartModelProducer.runTransaction` lambda based on the series _x_ range. Then, it would be read in `spacing`. If greater flexibility is required, a custom `HorizontalAxis.ItemPlacer` implementation—whether entirely standalone or based on delegation to the implementation instantiated by `HorizontalAxis.ItemPlacer.aligned`—can be employed instead. The `HorizontalAxis.ItemPlacer` functions are passed all of the required data; see, for example, the `fullXRange` parameter of `getLabelValues`.

   In Vico 2.0.0 Beta 3 and earlier, a custom `HorizontalAxis.ItemPlacer` (or `AxisItemPlacer.Horizontal`) implementation is universally required, as `spacing` isn’t an `ExtraStore` lambda but rather a plain integer.

   It’s also worth noting that, since version 2.0.0 Alpha 7, Vico offers dynamic `HorizontalAxis` label spacing, which automatically adjusts the label spacing to prevent truncation. This can replace manual spacing adjustment in some cases. In Vico 2.0.0 Beta 1 and later, dynamic label spacing applies when you use `HorizontalAxis.ItemPlacer.aligned` with `addExtremeLabelPadding` set to `true`. Similarly, in older versions, it applies when you use `HorizontalLayout.FullWidth` and `HorizontalAxis.ItemPlacer.default` (or `AxisItemPlacer.Horizontal.default`) with `addExtremeLabelPadding` set to `true`.

2. Since you’ve not specified otherwise, I assume you’re using `HorizontalLegend` or `VerticalLegend`. `Legend`s are applied directly, without the use of an `ExtraStore` lambda or provider interface. However, you can effectively hide a `Legend` by having it draw nothing and report a zero height.

   Since Vico 2.0.0 Alpha 28, the `items` properties of `HorizontalLegend` and `VerticalLegend` are `ExtraStore` lambdas. Thus, you can achieve the desired effect by means of an extra-based solution that removes all `LegendItem`s when only a single series is present. (See [here](https://github.com/patrykandpatrick/vico/blob/117a933a7ae515ea9682a1894d1eaf0f65d007c0/sample/src/main/java/com/patrykandpatrick/vico/sample/showcase/charts/DailyDigitalMediaUse.kt#L193) and [here](https://github.com/patrykandpatrick/vico/blob/117a933a7ae515ea9682a1894d1eaf0f65d007c0/sample/src/main/java/com/patrykandpatrick/vico/sample/showcase/charts/DailyDigitalMediaUse.kt#L101-L111) for an example of using extras with `HorizontalLegend`.) `HorizontalLegend`s with no `LegendItem`s are entirely invisible and take up no space, as desired. `VerticalLegend` is currently affected by an issue where, when no `LegendItem`s are present, space is still reserved for `VerticalLegend.padding`. This will be corrected. Should you be affected by this deficiency, you can resolve it by creating a `VerticalLegend` subclass and making the required adjustments.

   In Vico 2.0.0 Alpha 27 and earlier, a subclass or a custom `Legend` implementation is universally required, as the `items` properties aren’t `ExtraStore` lambdas but rather plain `LegendItem` lists.

   Regarding color adjustment, we’d need some more details to point you to a specific approach. If you’re referring to `LegendItem` colors, then the above applies, since `LegendItem`s are created in the `items` lambdas, as seen in the example provided. As for line colors, the optimal solution is heavily dependent on the specifics of the desired behavior. For more complex use cases, `LineCartesianLayer.LineProvider`, introduced in Vico 2.0.0 Alpha 23, may be required.

3. To adjust a `CartesianChart`’s _x_ and _y_ ranges based on the data’s ranges, one should create a custom implementation of `CartesianLayerRangeProvider` (for Vico 2.0.0 Beta 1; see [here](https://github.com/patrykandpatrick/vico/blob/bb10a34/sample/src/main/java/com/patrykandpatrick/vico/sample/showcase/charts/TemperatureAnomalies.kt#L58-L67) for an example) or `AxisValueOverrider` (for older versions). Its functions are passed the intrinsic ranges’ bounds. This is discussed in [subsection 2.3.2](https://www.patrykandpatrick.com/vico/guide/latest/core/cartesian-charts/cartesian-layer#232-cartesianlayerrangeprovider) of the guide.
